README.md
# Minehunter

[![Gem Version](https://badge.fury.io/rb/minehunter.svg)][gem]
[![Actions CI](https://github.com/piotrmurach/minehunter/workflows/CI/badge.svg?branch=master)][gh_actions_ci]
[![Build status](https://ci.appveyor.com/api/projects/status/6jnp11d5jpcvua4j?svg=true)][appveyor]
[![Maintainability](https://api.codeclimate.com/v1/badges/bce6981d523e678029e7/maintainability)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/github/piotrmurach/minehunter/badge.svg)][coverage]

[gem]: https://badge.fury.io/rb/minehunter
[gh_actions_ci]: https://github.com/piotrmurach/minehunter/actions?query=workflow%3ACI
[appveyor]: https://ci.appveyor.com/project/piotrmurach/minehunter
[codeclimate]: https://codeclimate.com/github/piotrmurach/minehunter/maintainability
[coverage]: https://coveralls.io/github/piotrmurach/minehunter

> Terminal mine hunting game built with [TTY toolkit components](https://ttytoolkit.org/components/).

Minehunter is a terminal puzzle game inspired by the classic "Minesweeper".

To win a game, a player needs to uncover all the fields inside the grid that don't contain mines. Uncovering a field with a mine immediately ends the game. When an uncovered field has no mine present, it can either be empty or show a number. The number specifies how many mines are in surrounding fields. Armed with this knowledge, the player can deduce fields safe to uncover and use a flag to mark the ones that hide a mine. The number of available flags matches the number of the randomly placed mines. But, the player is not required to use any flags to win the game.

There are three preconfigured difficulty levels:

* Easy - 9x9 with 10 mines
* Medium - 16x16 with 40 mines
* Hard - 32x32 with 99 mines

The player can also create a custom grid with an arbitrary number of mines.

Here is an example of playing on a 20x10 grid with 30 mines:

![Playing Minehunter](https://github.com/piotrmurach/minehunter/raw/master/assets/minehunter_custom_grid_play.png)

## Installation

Add this line to your application's Gemfile:

```ruby
gem "minehunter"
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install minehunter

## Usage

To start the game, run:

```shell
$ minehunter
```

Alternatively, run an alias:

```shell
$ minehunt
```

To change difficulty level use `--level` or `-l` option with `easy`, `medium` or `hard`:

```shell
$ minehunter --level easy
```

To customise the number of grid columns use  `--cols` or `-c` option. Likewise, specify  `--rows` or `-r` option to set the number of grid rows. For example, to play a game on a grid with `20` columns and `15` rows do:

```shell
$ minehunter --cols 20 --rows 15
```

Use `--mines` or `-m` option to set number of mines to be randomly placed on a grid:

```shell
$ minehunter --mines 25
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/piotrmurach/minehunter. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/piotrmurach/minehunter/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [GNU Affero General Public License v3.0](https://opensource.org/licenses/AGPL-3.0).

## Code of Conduct

Everyone interacting in the Minehunter project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/piotrmurach/minehunter/blob/master/CODE_OF_CONDUCT.md).

## Copyright

Copyright (c) 2021 Piotr Murach. See [LICENSE.txt](LICENSE.txt) for further details.


CHANGELOG.md
# Change log

## [v0.1.0] - 2021-10-10

* Initial release

[v0.1.0]: https://github.com/piotrmurach/minehunter/compare/eacd99c...v0.1.0


spec/spec_helper.rb
# frozen_string_literal: true

if ENV["COVERAGE"] == "true"
  require "simplecov"
  require "coveralls"

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ])

  SimpleCov.start do
    command_name "spec"
    add_filter "spec"
  end
end

require "minehunter"
require "stringio"

class StringIO
  def wait_readable(*)
    true
  end
end

RSpec.configure do |config|
  config.example_status_persistence_file_path = ".rspec_status"

  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.max_formatted_output_length = nil
    c.syntax = :expect
  end
end


spec/unit/cli_spec.rb
# frozen_string_literal: true

RSpec.describe Minehunter::CLI do
  let(:output) { StringIO.new("".dup, "w+") }
  let(:input) { StringIO.new("".dup, "w+") }
  let(:env) { {"TTY_TEST" => true} }

  it "prints a game on a 7x4 grid with 5 mines and quits" do
    cli = described_class.new
    input << "\n" << "q"
    input.rewind

    cli.run(%w[-c 7 -r 4 -m 5], input: input, output: output, color: true,
                                env: env, screen_width: 40, screen_height: 20)

    expect(output.string.inspect).to eq([
      "\e[?25l\e[2J",
      "\e[3;7H┌─────────────────────────┐",
      "\e[4;7H│      ,-*                \e[4;33H│",
      "\e[5;7H│     (_) Minehunter      \e[5;33H│",
      "\e[6;7H│                         \e[6;33H│",
      "\e[7;7H│ Movement                \e[7;33H│",
      "\e[8;7H│      [↑]        [w]     \e[8;33H│",
      "\e[9;7H│   [←][↓][→]  [a][s][d]  \e[9;33H│",
      "\e[10;7H│                         \e[10;33H│",
      "\e[11;7H│ Actions                 \e[11;33H│",
      "\e[12;7H│   Toggle Flag  f        \e[12;33H│",
      "\e[13;7H│   Uncover      space    \e[13;33H│",
      "\e[14;7H│   Restart      r        \e[14;33H│",
      "\e[15;7H│   Quit         q        \e[15;33H│",
      "\e[16;7H│                         \e[16;33H│",
      "\e[17;7H│ Press any key to start! \e[17;33H│",
      "\e[18;7H└─────────────────────────┘",
      "\e[2J",
      "\e[7;15H┌─────────┐",
      "\e[8;15H│ Flags 5 \e[8;25H│",
      "\e[9;15H├─────────┤",
      "\e[10;15H│ ░░░░░░░ \e[10;25H│",
      "\e[11;15H│ ░░░\e[42m░\e[0m░░░ \e[11;25H│",
      "\e[12;15H│ ░░░░░░░ \e[12;25H│",
      "\e[13;15H│ ░░░░░░░ \e[13;25H│",
      "\e[14;15H└─────────┘",
      "\e[2J\e[1;1H\e[?25h"
    ].join.inspect)
  end

  it "prints help information and exits" do
    cli = described_class.new

    expect {
      cli.run(%w[--help], input: input, output: output, screen_width: 40,
                          screen_height: 20)
    }.to raise_error(SystemExit) { |err| expect(err.status).to eq(0) }

    expect(output.string).to eq([
      "Usage: rspec [OPTIONS]",
      "",
      "Hunt down all the mines and uncover remaining fields",
      "",
      "Options:",
      "  -c, --cols INT    Set number of columns",
      "  -h, --help        Print usage",
      "  -l, --level NAME  Set difficulty level (permitted: easy,medium,hard)",
      "                    (default \"medium\")",
      "  -m, --mines INT   Set number of mines",
      "  -r, --rows INT    Set number of rows",
      "  -v, --version     Print version",
      "",
      "Examples:",
      "  To play the game on a 20x15 grid with 35 mines run",
      "",
      "  $ rspec -c 20 -r 15 -m 35\n"
    ].join("\n"))
  end

  it "prints version and exits" do
    cli = described_class.new

    expect {
      cli.run(%w[--version], input: input, output: output)
    }.to raise_error(SystemExit) { |err| expect(err.status).to eq(0) }

    expect(output.string).to eq("#{Minehunter::VERSION}\n")
  end

  it "prints errors when grid rows and columns aren't integers" do
    cli = described_class.new

    expect {
      cli.run(%w[--cols=a --rows=b], input: input, output: output)
    }.to raise_error(SystemExit) { |err| expect(err.status).to eq(1) }

    expect(output.string).to eq([
     "Errors:\n",
     "  1) Cannot convert value of `a` into 'int' type for '--cols' option\n",
     "  2) Cannot convert value of `b` into 'int' type for '--rows' option\n"
    ].join)
  end

  it "prints an error when the number of mines exceeds grid size and exits" do
    cli = described_class.new

    expect {
      cli.run(%w[-c 5 -r 5 -m 50], input: input, output: output)
    }.to raise_error(SystemExit) { |err| expect(err.status).to eq(1) }

    expect(output.string).to eq("Error: cannot have more mines than " \
                                "available fields\n")
  end

  it "prints an error for unknown difficulty level and exits" do
    cli = described_class.new

    expect {
      cli.run(%w[--level=unknown], input: input, output: output)
    }.to raise_error(SystemExit) { |err| expect(err.status).to eq(1) }

    expect(output.string).to eq("Error: unpermitted value `unknown` " \
                                "for '--level' option: choose from easy,\n "\
                                "      medium, hard\n")
  end
end


spec/unit/run_spec.rb
# frozen_string_literal: true

RSpec.describe Minehunter, "#run" do
  it "runs the game by invoking the command line interface" do
    cli = double(:cli, run: nil)
    allow(described_class::CLI).to receive(:new).and_return(cli)

    described_class.run

    expect(cli).to have_received(:run)
  end
end


lib/minehunter/game.rb
# frozen_string_literal: true

require "tty-box"
require "tty-cursor"
require "tty-reader"

require_relative "grid"
require_relative "intro"

module Minehunter
  # Responsible for playing mine hunting game
  #
  # @api public
  class Game
    # The keys to exit game
    #
    # @api private
    EXIT_KEYS = [?\C-x, "q", "\e"].freeze

    # The terminal cursor clearing and positioning
    #
    # @api public
    attr_reader :cursor

    # Create a Game instance
    #
    # @param [IO] input
    #   the input stream, defaults to stdin
    # @param [IO] output
    #   the output stream, defaults to stdout
    # @param [Hash] env
    #   the environment variables
    # @param [Integer] width
    #   the number of columns
    # @param [Integer] height
    #   the number of rows
    # @param [Integer] mines_limit
    #   the total number of mines
    # @param [Integer] screen_width
    #   the terminal screen width
    # @param [Integer] screen_height
    #   the terminal screen height
    # @param [Pastel] decorator
    #   the decorator for styling
    # @param [Proc] randomiser
    #   the random number generator
    #
    # @api public
    def initialize(input: $stdin, output: $stdout, env: {},
                   width: nil, height: nil, mines_limit: nil,
                   screen_width: nil, screen_height: nil,
                   decorator: DEFAULT_DECORATOR, randomiser: DEFAULT_RANDOMISER)
      @output = output
      @width = width
      @top = (screen_height - height - 4) / 2
      @left = (screen_width - width - 4) / 2
      @pos_x = (width - 1) / 2
      @pos_y = (height - 1) / 2
      @decorator = decorator
      @randomiser = randomiser
      @box = TTY::Box
      @cursor = TTY::Cursor
      @reader = TTY::Reader.new(input: input, output: output, env: env,
                                interrupt: :exit)
      @grid = Grid.new(width: width, height: height, mines_limit: mines_limit)
      @intro = Intro
      @intro_top = (screen_height - @intro.height - 2) / 2
      @intro_left = (screen_width - @intro.width - 4) / 2

      reset
    end

    # Reset game
    #
    # @api public
    def reset
      @curr_x = @pos_x
      @curr_y = @pos_y
      @first_uncover = true
      @lost = false
      @stop = false
      @grid.reset
    end

    # Check whether or not the game is finished
    #
    # @return [Boolean]
    #
    # @api public
    def finished?
      @lost || @grid.cleared?
    end

    # Start the game
    #
    # @api public
    def run
      @output.print cursor.hide + cursor.clear_screen + render_intro_box
      pressed_key = @reader.read_keypress
      keyctrl_x if EXIT_KEYS.include?(pressed_key)

      @output.print cursor.clear_screen
      @reader.subscribe(self)

      until @stop
        @output.print render_status_box + render_grid_box
        @reader.read_keypress
      end
    ensure
      @output.print cursor.show
    end

    # Render box with intro
    #
    # @return [String]
    #
    # @api private
    def render_intro_box
      @box.frame(
        @intro.render,
        top: @intro_top,
        left: @intro_left,
        padding: [0, 1]
      )
    end

    # Render box with status message
    #
    # @return [String]
    #
    # @api private
    def render_status_box
      @box.frame(
        status,
        top: @top,
        left: @left,
        width: @width + 4,
        padding: [0, 1],
        border: {bottom: false},
        align: :center
      )
    end

    # Render box with grid
    #
    # @return [String]
    #
    # @api private
    def render_grid_box
      @box.frame(
        render_grid,
        top: @top + 2,
        left: @left,
        padding: [0, 1],
        border: {
          top_left: :divider_right,
          top_right: :divider_left
        }
      )
    end

    # Status message
    #
    # @return [String]
    #
    # @api public
    def status
      if @lost
        "GAME OVER"
      elsif @grid.cleared?
        "YOU WIN"
      else
        "Flags #{@grid.flags_remaining}"
      end
    end

    # Render grid with current position marker
    #
    # @api private
    def render_grid
      @grid.render(@curr_x, @curr_y, decorator: @decorator)
    end

    # Control game movement and actions
    #
    # @param [TTY::Reader::KeyEvent] event
    #   the keypress event
    #
    # @api private
    def keypress(event)
      case event.value.to_sym
      when :h, :a then keyleft
      when :l, :d then keyright
      when :j, :s then keydown
      when :k, :w then keyup
      when :f, :g then flag
      when :r then reset
      when :q then keyctrl_x
      end
    end

    # Place a flag
    #
    # @api private
    def flag
      return if finished?

      @grid.flag(@curr_x, @curr_y) unless finished?
    end

    # Quit game
    #
    # @api private
    def keyctrl_x(*)
      @output.print cursor.clear_screen
      @output.print cursor.move_to(0, 0)
      @stop = true
    end
    alias keyescape keyctrl_x

    # Uncover a field
    #
    # @api private
    def keyspace(*)
      return if @grid.flag?(@curr_x, @curr_y)

      if @first_uncover
        @grid.fill_with_mines(@curr_x, @curr_y, randomiser: @randomiser)
        @first_uncover = false
      end
      @lost = @grid.uncover(@curr_x, @curr_y)
    end
    alias keyenter keyspace
    alias keyreturn keyspace

    # Move cursor up
    #
    # @api private
    def keyup(*)
      return if finished?

      @curr_y = @grid.move_up(@curr_y)
    end

    # Move cursor down
    #
    # @api private
    def keydown(*)
      return if finished?

      @curr_y = @grid.move_down(@curr_y)
    end

    # Move cursor left
    #
    # @api private
    def keyleft(*)
      return if finished?

      @curr_x = @grid.move_left(@curr_x)
    end

    # Move cursor right
    #
    # @api private
    def keyright(*)
      return if finished?

      @curr_x = @grid.move_right(@curr_x)
    end
  end # Game
end # Minehunter


lib/minehunter/grid.rb
# frozen_string_literal: true

require_relative "field"

module Minehunter
  # A grid with fields representation
  #
  # @api private
  class Grid
    # Track the number of flags remaining
    #
    # @return [Integer]
    #
    # @api public
    attr_reader :flags_remaining

    # Track the number of unmined fields remaining
    #
    # @return [Integer]
    #
    # @api public
    attr_reader :unmined_fields_remaining

    # Create a Grid instance
    #
    # @param [Integer] width
    #   the number of columns
    # @param [Integer] height
    #   the number of rows
    # @param [Integer] mines_limit
    #   the total number of mines
    #
    # @api public
    def initialize(width: nil, height: nil, mines_limit: nil)
      if mines_limit >= width * height
        raise Error, "cannot have more mines than available fields"
      end

      @width = width
      @height = height
      @mines_limit = mines_limit
      @fields = []

      reset
    end

    # Reset all fields to defaults
    #
    # @api public
    def reset
      (@width * @height).times do |i|
        @fields[i] = Field.new
      end
      @unmined_fields_remaining = @width * @height - @mines_limit
      @flags_remaining = @mines_limit
    end

    # Check whether or not the grid is cleared
    #
    # @return [Boolean]
    #
    # @api public
    def cleared?
      @unmined_fields_remaining.zero?
    end

    # All fields with mines
    #
    # @return [Array<Field>]
    #
    # @api public
    def mines
      @fields.select(&:mine?)
    end

    # Move up on the grid
    #
    # @return [Integer]
    #
    # @api public
    def move_up(y)
      y.zero? ? @height - 1 : y - 1
    end

    # Move down on the grid
    #
    # @return [Integer]
    #
    # @api public
    def move_down(y)
      y == @height - 1 ? 0 : y + 1
    end

    # Move left on the grid
    #
    # @return [Integer]
    #
    # @api public
    def move_left(x)
      x.zero? ? @width - 1 : x - 1
    end

    # Move right on the grid
    #
    # @return [Integer]
    #
    # @api public
    def move_right(x)
      x == @width - 1 ? 0 : x + 1
    end

    # Find field index at a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Integer]
    #
    # @api public
    def at(x, y)
      y * @width + x
    end

    # Find a field at a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Field]
    #
    # @api public
    def field_at(x, y)
      @fields[at(x, y)]
    end

    # Set a mine at a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @api public
    def mine(x, y)
      field_at(x, y).mine!
    end

    # Add or remove a flag at a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @api public
    def flag(x, y)
      field = field_at(x, y)
      return unless field.cover?

      @flags_remaining += field.flag? ? 1 : -1
      field.flag
    end

    # Check whether or not there is a flag at a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Boolean]
    #
    # @api public
    def flag?(x, y)
      field_at(x, y).flag?
    end

    # Fill grid with mines skipping the current position and nearby fields
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    # @param [Proc] randomiser
    #   the mine position randomiser
    #
    # @api public
    def fill_with_mines(x, y, randomiser: DEFAULT_RANDOMISER)
      limit = @mines_limit
      while limit > 0
        mine_x = randomiser[@width]
        mine_y = randomiser[@height]
        next if mine_x == x && mine_y == y
        next if fields_next_to(x, y).include?([mine_x, mine_y])

        field = field_at(mine_x, mine_y)
        next if field.mine?

        field.mine!
        limit -= 1
      end
    end

    # Enumerate fields next to a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Enumerator]
    #   the coordinates for nearby fields
    #
    # @api public
    def fields_next_to(x, y)
      return to_enum(:fields_next_to, x, y) unless block_given?

      -1.upto(1) do |offset_x|
        -1.upto(1) do |offset_y|
          close_x = x + offset_x
          close_y = y + offset_y

          next if close_x == x && close_y == y
          next unless within?(close_x, close_y)

          yield(close_x, close_y)
        end
      end
    end

    # Check whether coordinates are within the grid
    #
    # return [Boolean]
    #
    # @api public
    def within?(x, y)
      x >= 0 && x < @width && y >= 0 && y < @height
    end

    # Total number of mines next to a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Integer]
    #
    # @api public
    def count_mines_next_to(x, y)
      fields_next_to(x, y).reduce(0) do |acc, cords|
        acc + (field_at(*cords).mine? ? 1 : 0)
      end
    end

    # Total number of flags next to a given position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Integer]
    #
    # @api public
    def count_flags_next_to(x, y)
      fields_next_to(x, y).reduce(0) do |acc, cords|
        acc + (field_at(*cords).flag? ? 1 : 0)
      end
    end

    # Uncover fields surrounding the position
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Boolean]
    #   whether or not uncovered a mine
    #
    # @api public
    def uncover(x, y)
      field = field_at(x, y)

      if field.mine?
        field.uncover
        uncover_mines
        return true
      end

      return uncover_around(x, y) unless field.cover?

      mine_count = count_mines_next_to(x, y)
      field.mine_count = mine_count
      flag(x, y) if field.flag?
      field.uncover
      @unmined_fields_remaining -= 1

      if mine_count.zero?
        fields_next_to(x, y) do |close_x, close_y|
          close_field = field_at(close_x, close_y)
          next if !close_field.cover? || close_field.mine?

          uncover(close_x, close_y)
        end
      end
      false
    end

    # Uncover fields around numbered field matching flags count
    #
    # @param [Integer] x
    #   the x coordinate
    # @param [Integer] y
    #   the y coordinate
    #
    # @return [Boolean]
    #   whether or not uncovered a mine
    #
    # @api public
    def uncover_around(x, y)
      field = field_at(x, y)
      uncovered_mine = false

      if count_flags_next_to(x, y) != field.mine_count
        return uncovered_mine
      end

      fields_next_to(x, y) do |close_x, close_y|
        close_field = field_at(close_x, close_y)
        next if !close_field.cover? || close_field.flag?

        uncover(close_x, close_y)

        uncovered_mine = true if close_field.mine?
      end

      uncovered_mine
    end

    # Uncover all mines without a flag
    #
    # @api public
    def uncover_mines
      @fields.each do |field|
        if field.mine? && !field.flag? || field.flag? && !field.mine?
          field.wrong if field.flag?
          field.uncover
        end
      end
    end

    # Render grid
    #
    # @return [String]
    #
    # @api public
    def render(x, y, decorator: DEFAULT_DECORATOR)
      out = []

      @height.times do |field_y|
        @width.times do |field_x|
          field = field_at(field_x, field_y)
          rendered_field = field.render(decorator: decorator)

          if field_x == x && field_y == y && decorator
            bg_color = field.mine? && !field.cover? ? :on_red : :on_green
            rendered_field = decorator[rendered_field, bg_color]
          end

          out << rendered_field
        end
        out << "\n"
      end

      out.join
    end
  end # Grid
end # Minehunter


lib/minehunter/field.rb
# frozen_string_literal: true

module Minehunter
  # A field on a gird representation
  #
  # @api private
  class Field
    BOMB = "*"
    COVER = "░"
    EMPTY = " "
    FLAG = "F"
    WRONG = "X"

    # Mappings of mine counts to colour names
    MINE_COUNT_TO_COLOR = {
      1 => :cyan,
      2 => :green,
      3 => :red,
      4 => :blue,
      5 => :magenta,
      6 => :yellow,
      7 => :bright_cyan,
      8 => :bright_green
    }.freeze

    # The number of mines in nearby fields
    #
    # @api public
    attr_accessor :mine_count

    # Create a Field instance
    #
    # @api public
    def initialize
      @flag = false
      @mine = false
      @cover = true
      @wrong = false
      @mine_count = 0
    end

    # Toggle flag for a covered field
    #
    # @api public
    def flag
      return unless cover?

      @flag = !@flag
    end

    # Whether or not there is a flag placed
    #
    # @return [Boolean]
    #
    # @api public
    def flag?
      @flag
    end

    # Mark as having a mine
    #
    # @api public
    def mine!
      @mine = true
    end

    # Whether or not the field has mine
    #
    # @return [Boolean]
    #
    # @api public
    def mine?
      @mine
    end

    # Uncover this field
    #
    # @api public
    def uncover
      @cover = false
    end

    # Whether or not the field has cover
    #
    # @return [Boolean]
    #
    # @api public
    def cover?
      @cover
    end

    # Mark as having wrongly placed flag
    #
    # @api public
    def wrong
      @wrong = true
    end

    # Whether or not a flag is placed wrongly
    #
    # @return [Boolean]
    #
    # @api public
    def wrong?
      @wrong
    end

    # Render the field
    #
    # @param [Proc] decorator
    #   apply style formatting
    #
    # @return [String]
    #
    # @api public
    def render(decorator: DEFAULT_DECORATOR)
      if !cover?
        if mine? then BOMB
        elsif flag? && wrong? then decorator[WRONG, :on_red]
        elsif !mine_count.zero?
          decorator[mine_count.to_s, MINE_COUNT_TO_COLOR[mine_count]]
        else EMPTY end
      elsif flag? then FLAG
      else COVER end
    end
  end # Field
end # Minehunter


lib/minehunter/intro.rb
# frozen_string_literal: true

module Minehunter
  # An intro screen content
  #
  # @api private
  class Intro
    INTRO = [
      "     ,-*",
      "    (_) Minehunter",
      "",
      "Movement",
      "     [↑]        [w]",
      "  [←][↓][→]  [a][s][d]",
      "",
      "Actions",
      "  Toggle Flag  f",
      "  Uncover      space",
      "  Restart      r",
      "  Quit         q",
      "",
      "Press any key to start!"
    ].freeze

    # The maximum intro screen content width
    #
    # @return [Integer]
    #
    # @api public
    def self.width
      @width ||= INTRO.max_by(&:length).size
    end

    # The intro screen content height
    #
    # @return [Integer]
    #
    # @api public
    def self.height
      @height ||= INTRO.size
    end

    # Render intro screen content
    #
    # @return [String]
    #
    # @api public
    def self.render
      INTRO.join("\n")
    end
  end # SplashScreen
end # Minehunter


lib/minehunter/cli.rb
# frozen_string_literal: true

require "pastel"
require "tty-option"
require "tty-screen"

require_relative "game"
require_relative "version"

module Minehunter
  # The main interface to the game
  #
  # @api public
  class CLI
    include TTY::Option

    LEVELS = {
      "easy" => {width: 9, height: 9, mines: 10},
      "medium" => {width: 16, height: 16, mines: 40},
      "hard" => {width: 30, height: 16, mines: 99}
    }.freeze

    usage do
      no_command

      desc "Hunt down all the mines and uncover remaining fields"

      example "To play the game on a 20x15 grid with 35 mines run"

      example "$ #{program} -c 20 -r 15 -m 35"
    end

    option :width do
      short "-c"
      long "--cols INT"
      desc "Set number of columns"
      convert :int
    end

    option :height do
      short "-r"
      long "--rows INT"
      desc "Set number of rows"
      convert :int
    end

    option :level do
      default "medium"
      short "-l"
      long "--level NAME"
      desc "Set difficulty level"
      permit %w[easy medium hard]
    end

    option :mines do
      short "-m"
      long "--mines INT"
      desc "Set number of mines"
      convert :int
    end

    flag :help do
      short "-h"
      long "--help"
      desc "Print usage"
    end

    flag :version do
      short "-v"
      long "--version"
      desc "Print version"
    end

    # Run the game
    #
    # @param [Array<String>] argv
    #   the command line parameters
    # @param [IO] input
    #   the input stream, defaults to stdin
    # @param [IO] output
    #   the output stream, defaults to stdout
    # @param [Hash] env
    #   the environment variables
    # @param [Boolean] color
    #   whether or not to style the game
    # @param [Integer] screen_width
    #   the terminal screen width
    # @param [Integer] screen_height
    #   the terminal screen height
    #
    # @api public
    def run(argv = ARGV, input: $stdin, output: $stdout, env: {}, color: nil,
            screen_width: TTY::Screen.width, screen_height: TTY::Screen.height)
      parse(argv)

      if params[:help]
        output.print help
        exit
      elsif params[:version]
        output.puts VERSION
        exit
      elsif params.errors.any?
        output.puts params.errors.summary
        exit 1
      else
        level = LEVELS[params[:level]]
        decorator = Pastel.new(enabled: color).method(:decorate)
        game = Game.new(
          input: input,
          output: output,
          env: env,
          width: params[:width] || level[:width],
          height: params[:height] || level[:height],
          screen_width: screen_width,
          screen_height: screen_height,
          mines_limit: params[:mines] || level[:mines],
          decorator: decorator
        )
        game.run
      end
    rescue Minehunter::Error => err
      output.puts "Error: #{err}"
      exit 1
    end
  end # CLI
end # Minehunter
